### ISSUE TITLE
upload dataset with CLI stops working halfway through 

### ISSUE THREAD

**Issue by LinaTeichmann1** [OpenNeuroOrg/openneuro#2748]
Hi there,

I'm struggling with uploading MEG data to an existing OpenNeuro Dataset (ds004212). I'm running cli on a HPC system with:
`Kernel: Linux 3.10.0-1160.81.1.el7.x86_64
Operating System: Red Hat Enterprise Linux`
and `openneuro 4.14.1`

I am using the following command to upload the data:
openneuro upload -i -n --ignoreSubjectConsistency --dataset ds004212 BIDS-PATH

It passes the BIDS-validator and the upload begins but I'm getting multiple errors and it stalls at 71% with the following stdout:

```
Adding files to "ds004212"
bids-validator@1.9.9
Summary:                     Available Tasks:        Available Modalities: 
15294 Files, 237.69GB        main                    MEG                   
4 - Subjects                                         MRI                   
13 - Sessions       
```                                                    
```
Checking remote files...
(node:29443) ExperimentalWarning: The Fetch API is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
=======================================================================
2342 files to be uploaded with a total size of 128.1 GB
? Begin upload? Yes
=======================================================================
Starting a new upload (4d44d2c1) to dataset: 'ds004212'
ds004212 [----------------------------------------] 0% | ETA: 0s | 0/2342(node:29443) MaxListenersExceededWarning: Possible EventTarge
ds004212 [=============================-----------] 71% | ETA: 928s | 1681/2342Retrying upload for derivatives/preprocessed/eyes_epoched_cleaned_P1_S11.c
Retrying upload for derivatives/preprocessed/eyes_epoched_cleaned_P1_S11.csv: Error: Failed to upload file after 5 attempts - "https://openneuro.org/uplo
Retrying upload for derivatives/preprocessed/eyes_epoched_cleaned_P1_S6.csv: Bad Gateway
Retrying upload for derivatives/preprocessed/eyes_epoched_cleaned_P1_S6.csv: Error: Failed to upload file after 11 attempts - "https://openneuro.org/uplo
Retrying upload for derivatives/preprocessed/eyes_epoched_cleaned_P1_S3.csv: Bad Gateway
Retrying upload for derivatives/preprocessed/eyes_epoched_cleaned_P1_S3.csv: Error: Failed to upload file after 8 attempts - "https://openneuro.org/uploa
Retrying upload for derivatives/preprocessed/eyes_epoched_cleaned_P2_S1.csv: Bad Gateway
Retrying upload for derivatives/preprocessed/eyes_epoched_cleaned_P2_S1.csv: Error: Failed to upload file after 15 attempts - "https://openneuro.org/uplo
Retrying upload for derivatives/preprocessed/eyes_epoched_cleaned_P1_S7.csv: Bad Gateway
Retrying upload for derivatives/preprocessed/eyes_epoched_cleaned_P1_S7.csv: Error: Failed to upload file after 12 attempts - "https://openneuro.org/uplo
ds004212 [=============================-----------] 71% | ETA: 1005s | 1686/2342Killed
```

Any idea what might be going wrong? 

Thanks a lot, 
Lina

**Comment by nellh** [comment]
What version of Node.js are you using? This looks like you might be hitting a limit on the number of open file handles.

**Comment by nellh** [comment]
One more thing that would be helpful is if you can provide the file handle limit for the system you're uploading from. `ulimit -n` to obtain that value.

**Comment by LinaTeichmann1** [comment]
Sorry about that -- I'm using `node.js v18.12.1` and `ulimit -n` gives back `8192`

**Comment by LinaTeichmann1** [comment]
Hi @nellh, just following up on this: could you tell me what kind of memory usage is expected and if an open file descriptor limit of 8192 is likely a limitation? Thanks!

**Comment by nellh** [comment]
> Hi @nellh, just following up on this: could you tell me what kind of memory usage is expected and if an open file descriptor limit of 8192 is likely a limitation? Thanks!

Memory usage should be a few hundred megabytes and we target a minimum of 1024 file descriptors, so 8192 shouldn't be an issue. I'm still working on reproducing this to debug it. We recently switched to using the Node.js native fetch implementation for these requests and this might be related to that change.

You could try @openneuro/cli@4.12.1 which predates the change, that would be good to verify if it is the fetch changes causing this. It will give you a warning about being out of date but it is API compatible and should work. You will need to run it with `NODE_OPTIONS=--no-experimental-fetch` on Node 18.

**Comment by LinaTeichmann1** [comment]
Thanks for your reply @nellh! Can you tell me how I can change the NODE_OPTIONS to no-experiment-fetch? I can't see this as a variable in nodejs. Thanks!

**Comment by nellh** [comment]
> Thanks for your reply @nellh! Can you tell me how I can change the NODE_OPTIONS to no-experiment-fetch? I can't see this as a variable in nodejs. Thanks!

It's an environment variable. You can specify it before the command, like this: `NODE_OPTIONS=--no-experimental-fetch openneuro ...`

**Comment by LinaTeichmann1** [comment]
Thanks so much for all the help @nellh. It worked for me with openneuro v4.12.1 and the node-options set to --no-experimental-fetch!

**Comment by nellh** [comment]
The underlying issue here is fixed in 4.14.3, please reopen if you are having issues still on 4.14.3 and Node 18. Thanks for the help debugging this.

For large datasets you may need to raise Node's heap limit depending on how Node was installed. The Node option for this is called `max-old-space-size`, raising it to 4096 should be sufficient for very large datasets.
